aten/src/ATen/native/cpu/SortingKernel.cpp

```diff
@@ -139,30 +143,82 @@ static void parallel_sort1d_kernel(
 }
 #endif
 
+template< class RandomIt, class Compare>
+static void inline parallel_sort( RandomIt first, RandomIt last, Compare comp, std::function<void(RandomIt, RandomIt, Compare)> sort_f ) {
+  int64_t n = last - first;
+  const int64_t chunk_size = std::max(at::divup(n, at::get_num_threads()), at::divup(n, at::divup(n, 512)));
+  const int64_t num_threads = at::divup(n, chunk_size);
+
+  at::parallel_for(
+    0, n, chunk_size, [&](int64_t begin, int64_t end) {
+      sort_f(first + begin, first + end, comp);
+    }
+  );
+
+  if (num_threads > 1) {
+    const int64_t merge_threads = (num_threads - ((int) chunk_size * num_threads != n)) / 2;
+    at::parallel_for(
+      0, chunk_size * 2 * merge_threads, chunk_size * 2, [&](int64_t begin, int64_t end) {
+        std::inplace_merge(first + begin, first + begin + chunk_size, first + end, comp);
+      }
+    );
+
+    for (int i=0; i < merge_threads - 1; i++) {
+      std::inplace_merge(first, first + chunk_size * 2 * (i + 1), first + chunk_size * 2 * (i + 2), comp);
+    }
+
+    int last_tails = num_threads - 1 - merge_threads - (merge_threads - 1);
+    for (int i=0; i < last_tails; i++) {
+      std::inplace_merge(first, first + merge_threads * chunk_size * 2 + i * chunk_size, i == last_tails - 1 ? last : first + merge_threads * chunk_size * 2 + i * chunk_size, comp);
+    }
+  }
+}
+
 template <typename scalar_t, typename value_accessor_t, typename indices_accessor_t>
 static inline void sort_kernel_impl(const value_accessor_t& value_accessor,
             const indices_accessor_t& indices_accessor,
             int64_t dim_size, bool descending, bool stable) {
   auto composite_accessor = CompositeRandomAccessorCPU<
     value_accessor_t, indices_accessor_t
   >(value_accessor, indices_accessor);
+#ifdef __GNUC__
+  if (descending) {
+    if (stable) {
+      __gnu_parallel::stable_sort(composite_accessor, composite_accessor + dim_size,
+            KeyValueCompDesc<scalar_t>());
+    } else {
+      __gnu_parallel::sort(composite_accessor, composite_accessor + dim_size,
+            KeyValueCompDesc<scalar_t>());
+    }
+  } else {
+    if (stable) {
+      __gnu_parallel::stable_sort(composite_accessor, composite_accessor + dim_size,
+            KeyValueCompAsc<scalar_t>());
+    } else {
+      __gnu_parallel::sort(composite_accessor, composite_accessor + dim_size,
+            KeyValueCompAsc<scalar_t>());
+    }
+  }
+#else
+  using CompositeIt = CompositeRandomAccessorCPU<value_accessor_t, indices_accessor_t>;
   if (descending) {
     if (stable) {
-      std::stable_sort(composite_accessor, composite_accessor + dim_size,
-        KeyValueCompDesc<scalar_t>());
+      parallel_sort(composite_accessor, composite_accessor + dim_size,
+            KeyValueCompDesc<scalar_t>(), std::stable_sort<CompositeIt, KeyValueCompDesc<scalar_t>>);
```

Comment:
Since PyTorch now lives in c++17, can't we just use `std::sort` with some `std::execution::parallel_*_policy`?

Reply:
Sorry, @Skylion007 , stepped on your feet :)

Reply:
Hi I've tried this locally both within PyTorch and with a minimal reproducible example using C++17 but it seems to have no effect, which is why we've opted for __gnu_parallel::sort

Reply:
It looks like there might be a soft dependence on TBB https://stackoverflow.com/questions/51031060/are-c17-parallel-algorithms-implemented-already.

Reply:
Hi, thanks for the review. __gnu_prallel seems to be about 2x faster so I've put it as a default option for gcc but I've changed the backup option to the suggested method using std::execution

Reply:
This is surprising to me. Is it because the version of GCC is too old to have an implementation for the execution policy (ie falling back to serial implementation), or that there is some stdlib edge cases / overhead that is slowing things down?

Reply:
https://gcc.gnu.org/onlinedocs/libstdc++/manual/parallel_mode_using.html -- I actually cannot grep for `D_GLIBCXX_PARALLEL`.

Reply:
@Skylion007  there must be a different implementation or an overhead using stdlib because using the execution policy still provides a benefit, just not as much as __gnu_parallel

Reply:
@nikitaved My understanding was that if #include <parallel/algorithm>, we wouldn't need to add `D_GLIBCXX_PARALLEL`

Reply:
@Ryo-not-rio, yes, but it is only used for the GNU extensions, no?

Reply:
I'm a little hesitant in enabling parallelization for everything at once since this may potentially introduce bugs to unforseen areas